keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1182 @@
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Kubectl
+
+`kubectl drain` command will be changed to create a NodeMaintenance object after marking node
+unschedulable. We will also change the implementation to skip applications that support
+workload migration. At first, we will just check deployments with a positive
+`.spec.strategy.rollingUpdate.maxSurge` value. To make this extensible, we propose a new Surge API
+that could be checked before disrupting any workload (see below). If the cluster,
+doesn't support the NodeMaintenance API, kubectl will perform the node drain in a backwards
+compatible way.
+
+`kubectl uncordon` will be amended with `--node-maintenance` option. This can be used to delete the
+NodeMaintenance object together with marking the node as schedulable again. The default value of
+this flag will be `false`.
+
+### NodeMaintenance API
+
+NodeMaintenance objects serve as an intent to remove or migrate pods from a particular node. We will
+also support setting a reason for the maintenance.
+
+```golang
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    NodeName string
+    Reason string
+}
+
+type NodeMaintenanceStatus struct {
+    // number of pods this node maintenance is requesting to terminate
+    TargetedPods int32
+}
+```
+
+The NodeMaintenance API is being introduced with a minimal set of features, but it is expected that
+additional features will be added to the API in the future.
+
+The design of the API should not assume that only one node is selected for the node maintenance.
+And a Node to NodeMaintenance mapping should not be used or encouraged. `.spec.nodeName` field
+could be made optional if the requirements change in the future, in favor of `.spec.nodeSelector`.
+This could cause pods on multiple nodes to be marked for the maintenance at the same time, so even
+the NodeMaintenance `.status` should be made general enough.
+
+<<[UNRESOLVED NodeMaintenance ]>>
+
+- It is unclear whether we should allow marking the NodeMaintenances as finished and keeping the
+  old maintenances in the system for bookkeeping purposes.
+
+<<[/UNRESOLVED]>>
+
+### NodeMaintenance Controller
+
+Node maintenance controller will be introduced and added to `kube-controller-manager`. It will
+observe NodeMaintenance objects and set `TargetedByNodeMaintenance` condition on selected pods.
+The pods would be selected according to the node. Controllers that own these pods, would observe
+them and start a removal or migration from these nodes upon detecting this condition.
```

Comment:
The ordering of pod conditions might matter to some operators. Take an operator that updates on a per failure domain for example (eg: https://github.com/kubernetes/kubernetes/issues/114010). Ideally the operator would get a signal to migrate a subset of pods at the same time, so a migration away from affected nodes could be done strategically.

I can't think of a better messaging mechanism than conditions on pods however, so this may just need to be a design consideration of any controller/operator that is using this new condition.

Reply:
Thanks, we can take that into consideration, but should depend on it. API conventions specify that [conditions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties)

> This collection should be treated as a map with a key of type

The operator can also observe and parse the NodeMaintenance object as well and/or wait for the pods to be marked. 

We can keep the the discussion open, how to achieve the best messaging mechanism.

Reply:
We have moved from the pod conditions to the dedicated API: https://github.com/kubernetes/enhancements/pull/4565